### PR TITLE
Unify definitions of `next_function` and its variant parameter

### PR DESCRIPTION
**Pull request by greg7mdp**
This is in preparation for the changes for issue #698.

I have figured out a way to implement issue #698 with minimal code changes, but it requires changing a function type (and its variant parameter) that is used throughout our code. 

Because these types are defined in many locations, I wanted to have a first PR with no functional changes, just to make sure that the type is defined in a single location in our code. The benefit is that it should be easy to see that this PR, even though it has many changes, does not change any behavior.

### CODE REVIEW

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/chain/include/eosio/chain/transaction_metadata.hpp [hunk 1]
```diff
@@ -10,6 +10,12 @@ namespace boost { namespace asio {
 
 namespace eosio { namespace chain {
 
+template<typename T>
+using next_function_variant = std::variant<fc::exception_ptr, T, std::function<void()>>;
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/transaction_metadata.hpp:14]: The idea is that the third variant type will be a function which, when returned by the API call, will be posted to the http_plugin thread pool for finishing the execution of the API. So an API can use this feature or not.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 2]
```diff
@@ -2169,13 +2169,15 @@ void read_write::push_transaction(const read_write::push_transaction_params& par
...
+   auto wrapped_next = [=](const next_function_variant<read_write::push_transaction_results>& result) {
       if (std::holds_alternative<fc::exception_ptr>(result)) {
          const auto& e = std::get<fc::exception_ptr>(result);
          results->emplace_back( read_write::push_transaction_results{ transaction_id_type(), fc::mutable_variant_object( "error", e->to_detail_string() ) } );
-      } else {
+      } else if (std::holds_alternative<read_write::push_transaction_results>(result)) {
          const auto& r = std::get<read_write::push_transaction_results>(result);
          results->emplace_back( r );
+      } else {
+         assert(0);
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply plugins/chain_plugin/chain_plugin.cpp:2180]: Yes, exactly. I put the assert there just to make sure I wouldn't forget to add what's needed here.

File: libraries/chain/include/eosio/chain/transaction_metadata.hpp [hunk 3]
```diff
@@ -10,6 +10,12 @@ namespace boost { namespace asio {
...
+
+template<typename T>
+using next_function = std::function<void(const next_function_variant<T>&)>;
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/transaction_metadata.hpp:17]: Ha, I was not sure. I'll try `types.hpp`, indeed seems like a better choice, thanks for the suggestion.

File: plugins/chain_api_plugin/chain_api_plugin.cpp [hunk 4]
```diff
@@ -62,6 +62,32 @@ parse_params<chain_apis::read_only::get_transaction_status_params, http_params_t
 
 #define CHAIN_RO_CALL_WITH_400(call_name, http_response_code, params_type) CALL_WITH_400(chain, ro_api, chain_apis::read_only, call_name, http_response_code, params_type)
 
+template<class API, class PARAMS_PARSER, class HANDLER>
+static api_entry make_api_entry(http_plugin &_http_plugin, API& api, const char* api_name,
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply plugins/chain_api_plugin/chain_api_plugin.cpp]: done

File: libraries/chain/include/eosio/chain/transaction_metadata.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/transaction_metadata.hpp]: done, thanks!

**Review by heifner — APPROVED** [review]

File: plugins/chain_plugin/include/eosio/chain_plugin/trx_retry_db.hpp [hunk 5]
```diff
@@ -2,12 +2,10 @@
 #include <eosio/chain/types.hpp>
 #include <eosio/chain/block_state.hpp>
 #include <eosio/chain/trace.hpp>
+#include <eosio/chain/transaction_metadata.hpp>
```
Comment (linh2931) [inline plugins/chain_plugin/include/eosio/chain_plugin/trx_retry_db.hpp:5]: Do you still need this include?

Comment (greg7mdp) [reply plugins/chain_plugin/include/eosio/chain_plugin/trx_retry_db.hpp]: Good catch, you are right we don't need it, I have removed it.

**Review by linh2931 — APPROVED** [review]

**Pull request merged**